api/reduce

maxwilliams94 opened the pull request:
MPI_REDUCE wrapped by interface mplib20_reduce - basic (probably wrong) implementations for int_scalar and int_vec (rank 1)

Commit:
update gitignore

- build/ directory
- any compiled objects

Commit:
Initial attempt at wrapping MPI_REDUCE using mpilib20_reduce_m module

Current TODOs (in-file)
* Unravel how MPI_Datatype operates
* Overload for MPI_IN_PLACE?
* Testing

Commit:
add module procedures to interface mpilib20_reduce

Commit:
The root argument of mpilib20_reduce_<T>() generics is optional, allowing overiding of the default master rank within mpilib20_init_finalise.

AlexBuccheri commented on src/routines/generic_templates.md:
```diff
@@ -0,0 +1,23 @@
+# Generic Subroutine Templates
```
Is this necessary?

maxwilliams94 replied:
Probably right, this kind of thing can just stay local to me while the subroutines are generated

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+        module procedure mpilib20_reduce_int_vec
+    end interface
+    
+    public :: mpilib20_REDUCE
```
- [x] mpilib20_REDUCE -> mpilib20_reduce

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+    subroutine mpilib20_reduce_int_scalar(sendbuf, recvbuf, operation, mpi_env, root)
+        use mpi_bindings, only : MPI_REDUCE, MPI_Op, &
+            MPI_Datatype, MPI_INTEGER
+        use mpilib20_init_finalise, only : root_id
```
- [x] This is ok to make globally avoid at the top of the module

maxwilliams94 replied:
Make all of these imports global?

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+
+    !> MPI_REDUCE WRAPPER: INTEGER
+    subroutine mpilib20_reduce_int_scalar(sendbuf, recvbuf, operation, mpi_env, root)
+        use mpi_bindings, only : MPI_REDUCE, MPI_Op, &
```
MPI_REDUCE, MPI_INTEGER, MPI_Datatype, MPI_Op are ok to make available at the top of the module as we'll use multiple times in the module, and the scope's already well-controlled with private

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+        !> Variable to receive reduced set
+        type(MPI_Datatype),          intent(inout)   :: recvbuf
+        !> MPI Operation
+        type(MPI_Op),                intent(in)      :: operation
```
We don't want to pass the operation. This should be inferable from the input data type.

maxwilliams94 replied:
How would MPI_MAX and MPI_MIN be infered? Same with MPI_PRODUCT and MPI_SUM?

AlexBuccheri replied:
Good point. I think we could make operation optional and have default behaviour as sum. My feeling is the majority of use cases are sum. Would also be easy to change.

AlexBuccheri replied:
- [ ] make operation optional and have sum as default.

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+            MPI_Datatype, MPI_INTEGER
+        use mpilib20_init_finalise, only : root_id
+        !> Variable containing set to be sent
+        type(MPI_Datatype),          intent(in)      :: sendbuf
```
Send and receive buffers should be integers in this case - they're not tided to MPI types.

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+        !> Variable containing set to be sent
+        type(MPI_Datatype),          intent(in)      :: sendbuf
+        !> Variable to receive reduced set
+        type(MPI_Datatype),          intent(inout)   :: recvbuf
```
I'm guessing this wants to be intent(out) as it shouldn't come in with a value

maxwilliams94 replied:
For arrays, I assume intent(inout) is required as this will have to be preallocated?

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+        type(mpi_env_type),          intent(inout)   :: mpi_env
+        !> Rank of the process to receive the reduced set (override default)
+        integer, optional,           intent(in)      :: root
+        integer                                      :: use_root
```
Ah, I always have the same problem with renaming options. use root implies a logical. In this case root is predetermined by us, so I guess it should look more like:
```
mpilib20_reduce_int_scalar(..., process_id)

integer, intent(in), optional :: process_id
integer :: process

if (present(process_id)) then
 process = process_id
else 
 process = root_id
endif

call MPI_REDUCE(..., process,...)

```

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+            use_root = root_id
+        end if
+        !> Element count based-on receiving buffer size
+        count = size(recvbuf)
```
Add assertions to check the caller got it right.

- [x] call assert(size(sendbuf) == size(recvbuf), "size(sendbuf) /= size(recvbuf)")

AlexBuccheri replied:
I'd remove. size(recvbuf) is self-explanatory

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+        !> Element count based-on receiving buffer size
+        count = size(recvbuf)
+
+        call MPI_REDUCE(sendbuf, recvbuf, count, MPI_INTEGER, operation, root, mpi_env%comm, mpi_env%ierror)
```
- [x] count -> size(recvbuf)

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+    end subroutine mpilib20_reduce_int_scalar
+
+    !> MPI_REDUCE WRAPPER: INTEGER(:)
+    !> Rank 1 arrays don't require flattening
```
This is presumably for you, but can be removed when merged ;) 

In general, the routine docs should give more than the routine name. I'd copy them from the official docs.
```
Reduces values on all processes to a single value on process defined by process_id
```
One can also provide a basic example and distinguish how the call differs with passing root.
Thinking about this, root_id should be part of the MPI env because if the communicator does not contain all available processes, the root/master could differ from that of MPI_COMM_WORLD

maxwilliams94 replied:
Sure this all makes sense.

Agreed that `root_id` should probably go in the MPI env

AlexBuccheri commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -0,0 +1,83 @@
+!> Wrapping of the MPI_REDUCE methods
+!> TODOs
+!> Unravel how MPI_Datatype operates
+!> Overload for MPI_IN_PLACE?
```
I'm not entirely sure how `MPI_IN_PLACE` works. For a single communicator, it's a pointless operation. Once they're split, we'd need to consider its use

maxwilliams94 replied:
MPI_IN_PLACE is actually quite complicated to overload - `MPI_IN_PLACE` replaces sendbuf for the root process (indicating that the sendbuf and recvbuf are the same) BUT on non-root processes, there needs to be a variable passed via sendbuf. Not sure what this means for overloading for when sendbuf is MPI_IN_PLACE, probably would have MPI_IN_PLACE as an optional logical with logic to ensure correct filling of the (in this case) sendbuf and recvbuf variables for MPI_REDUCE.

https://stackoverflow.com/questions/17741574/in-place-mpi-reduce-crashes-with-openmpi

AlexBuccheri replied:
Ok, makes sense to ignore in the first instance and add an issue to address. I've never needed to use personally and only seen it recently

- [ ] Create issue

AlexBuccheri reviewed (commented):
Done the first pass. I guess you'd really like me to hurry up and finish the MPI testing such that you can start writing tests as you go

Commit:
PR changes

- remove generic_template file - useful locally if useful during development
- mpilib20_REDUCE interface -> mpilib20_reduce
- global imports moved to module scope
- recvbuf (receiving buffer) for integer case -> intent(out)
- use_rank (integer) renamed to process_id. Persistent rank is then renamed to process in calls.
- reduce verbosity when inferring dimension of buffers. Use size(recvbuf) rather than storing in `count` variable.
- add assert to check that sendbuf and recvbuf are equal dimension
- update todos at module docs (remaining types etc)
- update todos at subroutine level: usage examples

maxwilliams94 commented on src/routines/mpilib20_reduce.F90:
```diff
@@ -16,68 +17,73 @@ module mpilib20_reduce_m
+    !> Reduces values on all processes to a single value
+    !> on process defined by process_id
+    !> TODO usage examples
+    subroutine mpilib20_reduce_int_scalar(sendbuf, recvbuf, &
```
How do we want to line-split (to keep lines below 90 characters(?)). Here I've split after argument 2 because otherwise one goes over 90, but then operation (potentially being removed soon anyway) and mpi_env are grouped. I've put optional arguments on their own line

AlexBuccheri replied:
Yeah, not going over 90 (unless a comment) sounds sensible. I often like to put inputs and outputs on different lines but if we retain a consistent arg ordering with MPI commands, then clearly we can't. In which case do whatever looks neat. Not a fan of putting optionals on their own line.

More broadly we should introduce fprettify and set it up how we like, then we can just run it to apply consistent formatting

AlexBuccheri replied:
- [ ] 90 characters sounds good. Do what you think looks readable.

maxwilliams94 replied:
I'll keep it neat without putting optional arguments on their own line.

Sounds good re fprettify. We should get a config file written for it relatively soon. Once I commit these PR changes I'll have a good and see what get produced.

Commit:
Merge branch 'fix_proposed_structure' into api/REDUCE

Commit:
Remove non-existent file from src/CMakeLists.txt
